scribe bot cookbook. (#728)
elevenlabs/elevenlabs-docs: fern/docs/pages/cookbooks/speech-to-text/telegram-bot.mdx

---
title: Scribe Telegram Bot
subtitle: Build a Telegram bot that transcribes audio and video messages using TypeScript with Deno in Supabase Edge Functions.
---

## Introduction

In this tutorial you will learn how to build a Telegram bot that transcribes audio and video messages using TypeScript and the ElevenLabs Scribe model via the speech-to-text API.

<Tip title="Prefer to jump straight to the code?" icon="lightbulb">
  Find the [example project on
  GitHub](https://github.com/elevenlabs/elevenlabs-examples/tree/main/examples/speech-to-text/telegram-transcription-bot).
</Tip>

> ## Requirements

Reviewer: Also requires Deno, no?

Author: Yes, good catch, thanks! 🫶

// TODO
```

## Deploy the Edge Function

If you haven't already, create a new Supabase account at [database.new](https://database.new) and link the local project to your Supabase account:

```bash
supabase link
```

### Apply the database migrations

Run the following command to apply the database migrations from the `supabase/migrations` directory:

```bash
supabase db push
```

Navigate to the [table editor](https://supabase.com/dashboard/project/_/editor) in your Supabase dashboard and you should see and empty `transcription_logs` table.

![Empty table](/assets/images/cookbooks/scribe/telegram-bot/supa-empty-table.png)

Lastly, run the following command to deploy the Edge Function:

```bash
supabase functions deploy --no-verify-jwt scribe-bot
```

Navigate to the [Edge Functions view](https://supabase.com/dashboard/project/_/functions) in your Supabase dashboard and you should see the `scribe-bot` function deployed. Make a note of the function URL as you'll need it later, it should look something like `https://<project-ref>.functions.supabase.co/scribe-bot`.

![Edge Function deployed](/assets/images/cookbooks/scribe/telegram-bot/supa-edge-function-deployed.png)

> ## Register the Telegram bot

Reviewer: This and the webhook part should happen first so you can set up your `.env` file in one go.

Author: Yah, good point, thanks.

Next, use [the BotFather](https://t.me/BotFather) to create a new Telegram bot. Run the `/newbot` command and follow the instructions to create a new bot. At the end, you will receive your secret bot token. Set it in your `.env` file as `TELEGRAM_BOT_TOKEN`.

![BotFather](/assets/images/cookbooks/scribe/telegram-bot/bot-father.png)

### Set up the webhook

Set your bot's webhook url to `https://<PROJECT_REFERENCE>.functions.supabase.co/telegram-bot` (Replacing `<...>` with respective values). In order to do that, simply run a GET request to the following url (in your browser, for example):

```
https://api.telegram.org/bot<TELEGRAM_BOT_TOKEN>/setWebhook?url=https://<PROJECT_REFERENCE>.supabase.co/functions/v1/scribe-bot?secret=<FUNCTION_SECRET>
```

Note that the `FUNCTION_SECRET` is the secret you set in your `.env` file.

![Set webhook](/assets/images/cookbooks/scribe/telegram-bot/set-webhook.png)

## Set the function secrets

Now that you have all your secrets set locally, you can run the following command to set the secrets in your Supabase project:

```bash
supabase secrets set --env-file supabase/functions/.env
```

## Test the bot

> That's it, now you can test the bot by sending a voice message, audio, or video to your bot. Go on, try it out!

Reviewer: Need to revise the tone, docs and public comms shouldn't use exclamation marks.

```suggestion
Finally you can test the bot by sending it a voice message, audio or video file.
```

import 'jsr:@supabase/functions-js/edge-runtime.d.ts';
import { createClient } from 'jsr:@supabase/supabase-js@2';
import { ElevenLabsClient } from 'npm:elevenlabs@1.50.5';

console.log(`Function "elevenlabs-scribe-bot" up and running!`);

const elevenLabsClient = new ElevenLabsClient({
  apiKey: Deno.env.get('ELEVENLABS_API_KEY') || '',
});

const supabase = createClient(
  Deno.env.get('SUPABASE_URL') || '',
  Deno.env.get('SUPABASE_SERVICE_ROLE_KEY') || ''
);

async function scribe({
  fileURL,
  fileType,
  duration,
  chatId,
  messageId,
  username,
}: {
  fileURL: string;
  fileType: string;
  duration: number;
  chatId: number;
  messageId: number;
  username: string;
}) {
  let transcript: string | null = null;
  let languageCode: string | null = null;
  let errorMsg: string | null = null;
  try {
    const sourceFileArrayBuffer = await fetch(fileURL).then((res) => res.arrayBuffer());
    const sourceBlob = new Blob([sourceFileArrayBuffer], {
      type: fileType,
    });

    const scribeResult = await elevenLabsClient.speechToText.convert({
      file: sourceBlob,
      model_id: 'scribe_v1',
      tag_audio_events: false,
    });

>     // console.log({ scribeResult });

Reviewer: Can probably remove this.

    transcript = scribeResult.text;
    languageCode = scribeResult.language_code;

    // Reply to the user with the transcript
    await bot.api.sendMessage(chatId, transcript, {
      reply_parameters: { message_id: messageId },
    });
  } catch (error) {
    errorMsg = error.message;
    console.log(errorMsg);
    await bot.api.sendMessage(chatId, 'Sorry, there was an error. Please try again.', {
      reply_parameters: { message_id: messageId },
    });
  }
  // Write log to Supabase.
  const logLine = {
    file_type: fileType,
    duration,
    chat_id: chatId,
    message_id: messageId,
    username,
    language_code: languageCode,
    error: errorMsg,
  };
  console.log({ logLine });
  await supabase.from('transcription_logs').insert({ ...logLine, transcript });
}

const telegramBotToken = Deno.env.get('TELEGRAM_BOT_TOKEN');
const bot = new Bot(telegramBotToken || '');

> const startMessage = `Welcome to the ElevenLabs Scribe Bot\\! I can transcribe speech from 80 different languages with super high accuracy\\! 

Reviewer: Nit, but is the number 80 languages? Not sure what the final tally was.

Reviewer 2: Will have a final number by EOD

Author: Should we go with `I can transcribe speech in 80+ languages with super high accuracy`?
